imports/lib/schemas/customTypes.ts

```diff
@@ -0,0 +1,43 @@
+import { Meteor } from 'meteor/meteor';
+import { z } from 'zod';
+import { Id } from './regexes';
+
+// Each of these is set to true based on the context in which the schema is
+// being evaluated. They are mutually exclusive (although they can all be false)
+export const IsInsert = new Meteor.EnvironmentVariable<boolean>();
+export const IsUpdate = new Meteor.EnvironmentVariable<boolean>();
+export const IsUpsert = new Meteor.EnvironmentVariable<boolean>();
+
+// In several of these auto-generated fields, we lie to zod about the output
+// type of the field - with type assertions, we say that they are non-nullable.
+// This will be true in the database, even though the transform functions don't
+// always return values (e.g. createdTimestamp will always be a Date, it'll just
+// get set on insertion and not on subsequent updates).
+
+export const stringId = z.string().regex(Id).optional()
+  .transform((v) => v as unknown as string);
```

Comment:
Do we want to generate the `Random.id()` here if it wasn't provided, or just continue delegating that to Meteor?

Reply:
I figured it was easiest to continue delegating to Meteor. Meteor has some complex behavior around ID generation in order to ensure that (e.g.) stub method implementations generate the same IDs as their corresponding server-side runs, and even if we're not relying on that behavior, I didn't see any reason to suppress it.